Schema validation api v2 (#813)
liveblocks/liveblocks: docs/pages/guides/schema-validation.mdx

should only happen in development mode, and the developer is responsible for
fixing them.

> ## Schema Operations

Reviewer: ```suggestion
## Schema actions
```

I'd prefer not to use "operations" here, because other things are called operations already.

You can interact with the schemas you create through the dashboard or by calling

> the [rest API](/docs/api-reference/rest-api-endpoints). In this guide, we will

Reviewer: REST is an acronym, so we should use uppercase here:

```suggestion
the [REST API](/docs/api-reference/rest-api-endpoints). In this guide, we will
```

cover the following operations:

- [Create a schema](#Create-a-schema)
- [Delete a schema](#Delete-a-schema)
- [Attach a schema to a room](#Attach-a-schema-to-a-room)

> - [Detach a schema from a room](#Detach-a-schema-from-a-room)

Reviewer: More of a style/consistency thing, not super important. But in other guides, our section titles are in present progressive tense form, e.g. "Creat**ing** a schema", "Detach**ing** a schema from a room" etc. Perhaps keep those consistent? No strong opinion here. Perhaps we don't do it consistently yet, in which case, feel free to disregard.

> You can also use the rest API to

Reviewer: Ditto.

[list schemas](/docs/api-reference/rest-api-endpoints#list-schemas),
[get a schema](/docs/api-reference/rest-api-endpoints#get-schema), and
[update a schema](/docs/api-reference/rest-api-endpoints#update-schema).

### Create a schema

Creating a schema via the dashboard is straightforward. Simply navigate to the

> “Schemas” page in the project you want to add the schema, and click on the

Reviewer: Heads up: latest thinking is that we’ll remove “Schemas” from the top level nav and eventually make it more of a secondary citizen in the dashboard (accessible via the rooms list/detail pages). Steven and Pierre have some ideas/designs for that.

For now though, absolutely true. Just a reminder that we should update these instructions when we change the UI.

Author: thanks for the heads up :)

  </video>
</Figure>

Another way to create a schema is via the
[create schema API](/docs/api-reference/rest-api-endpoints#post-create-new-schema).

> When a schema is sent to the API it needs to be constructed as a string literal
> to be valid JSON when sent to the API for parsing. In the dashboard, a sample
> schema could be represented as:

Reviewer: The schema body needs no special treatment and we shouldn't put too much emphasis on it. The trouble you ran into came mostly from the fact that you tried to use cURL, and newline characters give problems in a command line environment and in raw JSON, in case you would write that.

Typically, people will send these requests from JavaScript though, in which case `JSON.stringify()` will handle strings with newlines just like any other string, and it won't be an issue.

That said, when you are explaining how to make a cURL request, then it may be useful to mention this. In this case, it's not needed.

Suggestion:

```suggestion
A schema body is a multi-line string using the Liveblocks <link>schema syntax</link>. In the dashboard, a sample
schema could be represented as:
```

```ts
type Storage {
  todos: LiveList<LiveObject<Todo>>
}

type Todo {
  text: string
  checked?: boolean
}
```

To convert the schema to a string literal which will be used in the `body` key
of the request, you can use the following code:

```ts
JSON.stringify({
  body: `
type Storage {
  todos: LiveList<LiveObject<Todo>>
}

type Todo {
  text: string
  checked?: boolean
}
`,
}).replace(/\\n/g, "\n");
// returns '{"body":"\ntype Storage {\n  todos: LiveList<LiveObject<Todo>>\n}\n\ntype Todo {\n  text: string\n  checked?: boolean\n}\n"}'

> ```

Reviewer: All of this is not needed, and can be removed.

Author: my beautiful hack 😭 🤣

Reviewer: Gotta kill your darlings sometimes 😞 Sorry if I made things confusing along the way somewhere!

To create the schema:

```ts
fetch("https://api.liveblocks.io/v2/schemas", {
  method: "POST",
  body: JSON.stringify({
    name: "todo-list",

>     body: "\ntype Storage {\n  todos: LiveList<LiveObject<Todo>>\n}\n\ntype Todo {\n  text: string\n  checked?: boolean\n}\n",

Reviewer: You can just inline the multi-line `` ` ``-enclosed string from the previous example here.

  }),
});
```

<Banner title="Schema syntax">

Learn more about our [schema syntax](/docs/guides/schema-validation/syntax) and
what kind of validations are currently supported.

</Banner>

### Delete a schema

In the dashboard, you can delete a schema by navigating to the schema in
question and clicking on the trash icon next to the schema version you want to
delete. If you have attached a schema to a room, you will need to detach the
schema from the room before you can delete it. If this schema is frozen, the
trash icon will be greyed out.

<Figure>
  <video autoPlay loop muted playsInline>
    <source
      src="/images/docs/schema-validation/delete-schema.mp4"
      type="video/mp4"
    />
  </video>
</Figure>

If your schema is in a detached state, you can also delete it by using the
[delete schema API](/docs/api-reference/rest-api-endpoints#delete-a-schema).

```ts
fetch("https://api.liveblocks.io/v2/schemas/{id}", {

>   method: "POST",

Reviewer: ```suggestion
  method: "DELETE",
```

});
```

### Attach a schema to a room

You can attach a schema to a room via the dashboard by navigating to the room in
question and clicking on the “Attach schema” button.

<Figure>
  <video autoPlay loop muted playsInline>
    <source
      src="/images/docs/schema-validation/attach-schema.mp4"
      type="video/mp4"
    />
  </video>
</Figure>

Alternatively, you can use the
[attach schema API](/docs/api-reference/rest-api-endpoints#post-attach-schema-to-room).
For example:

```ts
fetch("https://api.liveblocks.io/v2/rooms/{roomId}/schema", {
  method: "POST",
  body: JSON.stringify({
    schema: "todo-list@1",
  }),
});
```

### Detach a schema from a room

You can use the dashboard to remove a schema from a room by navigating to the
room in question and clicking on the dropdown menu which displays the currently
attached schema. From there, you can click on the “Detach schema” button.

> {" "}

Reviewer: ```suggestion
```

Author: prettier added that.  not sure why.  will try to remove
